### ISSUE TITLE
Data race in louvain_enactor.cuh

### ISSUE THREAD

**Issue by AKKamath** [gunrock/gunrock#828]
I found a data race in gunrock/app/louvain/louvain_enactor.cuh.
 
In line 469 community_sizes[v] and community_sizes[n_comm] are being read, while in lines 482 and 483, atomic operations are performed on community_sizes, creating a data race.

           if (c_comm == v &&
    -->         community_sizes[v] + community_sizes[n_comm] <= 2) {
              if (n_comm > v && next_communities[n_comm] == v) {
                max_gains[v] = 0;
                return;
              }
            }
            if (c_comm == n_comm) {
              return;
              max_gains[v] = 0;
            }
            // printf("v %d : c %d -> c %d, gain = %lf\n",
            //    v, c_comm, n_comm, max_gains[v]);
        
    -->    atomicSub(community_sizes + c_comm, 1);
    -->    atomicAdd(community_sizes + n_comm, 1);
            auto w_v2v = w_v2[v];
            atomicAdd(w_c2 + n_comm, w_v2v);
            atomicAdd(w_c2 + c_comm, -1 * w_v2v);
            communities[v] = n_comm;
          }

**Comment by jowens** [comment]
@AKKamath do you have an automated race detector or have you run this code and you found a race and you're pointing it out to us? It's something we should know about either way, but if you've seen a race in practice, that is helpful info for us.

**Comment by AKKamath** [comment]
This data race manifested when I ran the louvain application in the example folder with the provided inputs.

**Comment by sgpyc** [comment]
Louvain has a serial CPU reference, and the result is checked against the output of the CPU reference.
https://github.com/gunrock/gunrock/blob/master/examples/louvain/test_louvain.cu#L76
So, if the result with race conditions still matches the CPU reference, it might be that different vertex movement orders settle to the same results.

I roughly remembered the first if statement
`       if (c_comm == v &&
-->         community_sizes[v] + community_sizes[n_comm] <= 2) {
          if (n_comm > v && next_communities[n_comm] == v) {
            max_gains[v] = 0;
            return;
          }
        }
`
is to dis-allow vertex A move to vertex B's community, and vertex B move to vertex A's community at the same iteration, when both A&B's communities only have themselves. The 'next_communities[n_comm] == v' is a very strong condition, that  prevents most harm from the race. Even if the race happens, community_sizes[v] is at least 1 (v is still inside), so community_sizes[n_comm] is at most 1 when the condition is judged; if n_comm happens to be moving to v (when the inside if statement is true), v won't move to n_comm. I think this logic doesn't care whether community_sizes[n_comm] has been updated or not. Let's see the different situations:
0. c_s[n_comm] read 0, actual 0: good
1. c_s[n_comm] read 0, actual 1: some new vertex joined n_comm; if v moves wrongly, it can correct in next iteration.
2. c_s[n_comm] read 0, actual >1: same as above, but with more than one vertices joined n_comm.
3. c_s[n_comm] read 1, actual 0: v may moves to an empty community; it's okay (will join back if that's not optimal).
4. c_s[n_comm] read 1, actual 1: good
5. c_s[n_comm] read 1, actual >1: n_comm grows; if v not joining this iteration, it can join in next.
6. c_s[n_comm] read >1, actual 0: v may moves to an empty community; it's okay (will join back if that's not optimal).
7. c_s[n_comm] read >1, actual 1: n_comm reduces; if v joins wrongly, it can move out in next iteration.
8. c_s[n_comm] read >1, actual >1: good
It's quite clear that if the decision is wrong because of data race (although it should be very rare), it can be corrected in following iterations.

If the race really becomes an issue, it can be mitigate by using an old_community_sizes array to hold the original values before going into the kernel.

**Comment by jowens** [comment]
@sgpyc : so, legit race detected, application logic makes it non-harmful in practice?

Thank you for this analysis! This is a ~similar conclusion to https://github.com/gunrock/gunrock/issues/829.

**Comment by sgpyc** [comment]
Yep. The iterative updates can mitigate the unwanted effects of data races. PR is another algorithm that has this property, and the Gunrock implementation uses rank_curr & rank_next to avoid data races. I remembered someone did a correct implementation of PR that allows data races.

If the application logic was not sufficient to mitigate the data race risk, these might happen: 1) infinite loop, particularly A->B & B->A, or A->B, B->C, C->A, this kind of circular movements; or 2) diverged results with the serial CPU implementation.
